[WIP] Support v1alpha1.ScyllaDBDatacenter registration with global ScyllaDB Manager instance

rzetelskik opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/scylladb/scylla-operator/blob/master/CONTRIBUTING.md)
documentation before submitting a Pull Request!
Thank you for contributing to the Scylla Operator! -->

**Description of your changes:** wip

**Which issue is resolved by this Pull Request:**
Resolves #2570 

/kind feature
/priority important-soon
/cc

rzetelskik commented:
> @rzetelskik: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/e2e-gke-parallel	[f171058](https://github.com/scylladb/scylla-operator/commit/f17105821c98e6b09065732b3ff2bfc2c64b7215)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel/1911797166498648064)	true	`/test e2e-gke-parallel`
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).

/retest

rzetelskik commented:
> @rzetelskik: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name | Commit | Details | Required | Rerun command
> --- | --- | --- | --- | ---
> ci/prow/e2e-gke-parallel | f17105821c98e6b09065732b3ff2bfc2c64b7215 | [link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel/1911797166498648064) | true | `/test e2e-gke-parallel`
> ci/prow/e2e-gke-multi-datacenter-parallel | f17105821c98e6b09065732b3ff2bfc2c64b7215 | [link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-multi-datacenter-parallel/1911797169468215296) | true | `/test e2e-gke-multi-datacenter-parallel`
> ci/prow/e2e-gke-parallel-clusterip | f17105821c98e6b09065732b3ff2bfc2c64b7215 | [link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel-clusterip/1911797167501086720) | true | `/test e2e-gke-parallel-clusterip`
> 
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).
> 
> <details>
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
> </details>
> <!-- test report -->

/retest

rzetelskik commented:
> @rzetelskik: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name | Commit | Details | Required | Rerun command
> --- | --- | --- | --- | ---
> ci/prow/e2e-gke-multi-datacenter-parallel | f17105821c98e6b09065732b3ff2bfc2c64b7215 | [link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-multi-datacenter-parallel/1911827300089860096) | true | `/test e2e-gke-multi-datacenter-parallel`
> ci/prow/e2e-gke-parallel-clusterip | f17105821c98e6b09065732b3ff2bfc2c64b7215 | [link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel-clusterip/1911827299943059456) | true | `/test e2e-gke-parallel-clusterip`
> 
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).
> 
> <details>
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
> </details>
> <!-- test report -->

/test images
/retest

rzetelskik commented:
/cc zimnx mflendrich
Ready to review. Please pay special attention to differences between this implementation and the existing manager-controller.

rzetelskik commented:
@zimnx one thing I'm unsure about and so I'd like to hear your opinion. This PR introduces verification utils for `v1alpha1.ScyllaDBDatacenters`, most importantly the `Verify` function which is supposed to mirror the existing logic for `v1.ScyllaClusters`. The existing one checks for the upgrade status (https://github.com/scylladb/scylla-operator/blob/64e0841d44f800d366f8615517f863f1d9ce706d/test/e2e/utils/verification/scyllacluster/verify.go#L316), which, for ScyllaDBDatacenters is reflected in a conditionally existing ConfigMap. Should the upgrade context CM be verified as part of this method then? To me it seems like only verifying it in upgrade-related tests would suffice.

rzetelskik commented:
> @rzetelskik: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/e2e-gke-parallel	[d485a02](https://github.com/scylladb/scylla-operator/commit/d485a021338143bb30e3f35eeab7af5cae1dca0b)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel/1912117323142008832)	true	`/test e2e-gke-parallel`
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).

https://github.com/scylladb/scylla-operator/issues/2096#issuecomment-2805011187
known flake
/retest

zimnx commented:
> @zimnx one thing I'm unsure about and so I'd like to hear your opinion. This PR introduces verification utils for `v1alpha1.ScyllaDBDatacenters`, most importantly the `Verify` function which is supposed to mirror the existing logic for `v1.ScyllaClusters`. The existing one checks for the upgrade status (
> 
> https://github.com/scylladb/scylla-operator/blob/64e0841d44f800d366f8615517f863f1d9ce706d/test/e2e/utils/verification/scyllacluster/verify.go#L316
> ), which, for ScyllaDBDatacenters is reflected in a conditionally existing ConfigMap. Should the upgrade context CM be verified as part of this method then? To me it seems like only verifying it in upgrade-related tests would suffice.

Verify is executed post rollout. It should validate whether main and dependant objects and status are in expected state. 
Upgrade context ConfigMap shouldn't exist at all at this point. We could validate it's not existing, although it's a implementation detail so maybe black box approach would be better in the long term. I would prefer the latter.

rzetelskik commented:
- Modified GlobalScyllaDBManager controller to unit test `makeScyllaDBManagerClusterRegistrationForScyllaDBDatacenter` func

rzetelskik commented:
- Updated condition names to reflect sync func name

mflendrich commented on pkg/api/scylla/validation/scylladbmanagerclusterregistration_validation.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func ValidateScyllaDBManagerClusterRegistrationSpec(spec *scyllav1alpha1.ScyllaDBManagerClusterRegistrationSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
```
nit

note https://github.com/scylladb/scylla-operator/pull/2598 - this review inspired me to make that suggestion, please decide & get this PR in sync with that proposal.

rzetelskik replied:
done

zimnx commented on pkg/controller/scylladbmanagerclusterregistration/controller.go:
```diff
@@ -0,0 +1,356 @@
+		old.(*corev1.Namespace),
+		cur.(*corev1.Namespace),
+		smcrc.handlers.EnqueueAllWithUntypedFilterFunc(hasGlobalScyllaDBManagerNamespaceName),
+		smcrc.deleteNamespace,
```
EnqueueAll enqueues all ScyllaDBManagerClusterRegistration objects from untypedObj namespace. But this is Namespace so it has empty .Namespace field, which value turns out to be equal to special corev1.NamespaceAll, meaninig it will enqueue all ScyllaDBManagerClusterRegistration from all namespaces. Is this expected behavior?

rzetelskik replied:
Yes, e.g. deletion of `scylla-manager` namespace should unblock finalization of all existing SMCRs having the global-manager label. Although I can write a dedicated method if you think the route to this behaviour is hacky

zimnx replied:
it's very much hacky, lets have dedicated method expressing what is desired behavior

rzetelskik replied:
done

zimnx reviewed (commented):
first batch

rzetelskik commented on pkg/controller/globalscylladbmanager/sync_scylladbmanagerclusterregistrations.go:
```diff
@@ -0,0 +1,107 @@
+	var errs []error
+	for _, smcr := range requiredScyllaDBManagerClusterRegistrations {
+		_, _, err = resourceapply.ApplyScyllaDBManagerClusterRegistration(ctx, gsmc.scyllaClient, gsmc.scyllaDBManagerClusterRegistrationLister, gsmc.EventRecorder(), smcr, resourceapply.ApplyOptions{
+			AllowMissingControllerRef: true,
```
there is no object to set the progressing condition on, best we can do here is a log? I left it out as the method creates an event

zimnx replied:
ok

rzetelskik commented on pkg/controller/globalscylladbmanager/sync.go:
```diff
@@ -0,0 +1,65 @@
+
+func (gsmc *Controller) getScyllaDBManagerClusterRegistrations() (map[string]*scyllav1alpha1.ScyllaDBManagerClusterRegistration, error) {
+	smcrs, err := gsmc.scyllaDBManagerClusterRegistrationLister.List(labels.SelectorFromSet(labels.Set{
+		naming.GlobalScyllaDBManagerLabel: naming.LabelValueTrue,
```
Why? Both end up calling `ListAll`, it's just syntax sugar?

rzetelskik commented on pkg/controller/globalscylladbmanager/sync.go:
```diff
@@ -0,0 +1,65 @@
+	globalScyllaDBManagerSelector = labels.SelectorFromSet(labels.Set{
+		naming.GlobalScyllaDBManagerRegistrationLabel: naming.LabelValueTrue,
+	})
+)
```
> we keep selectors in naming package

only some of them, see e.g.
https://github.com/scylladb/scylla-operator/blob/e6910a520be1c15ef266f5846d6a6c782224a160/pkg/controller/nodeconfig/sync.go#L53
https://github.com/scylladb/scylla-operator/blob/e6910a520be1c15ef266f5846d6a6c782224a160/pkg/controller/scyllacluster/sync.go#L55

Is there a benefit to making it public?

rzetelskik replied:
I left this one as internal, but moved SMCR selector and selector labels to naming package as they are shared between controllers

rzetelskik commented on pkg/controller/globalscylladbmanager/sync_scylladbmanagerclusterregistrations.go:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	err = gsmc.pruneScyllaDBManagerClusterRegistrations(
+		ctx,
```
it's namespaced, but I'll fix it as per your comment here https://github.com/scylladb/scylla-operator/pull/2590#discussion_r2048694422

rzetelskik replied:
done

rzetelskik commented on pkg/api/scylla/v1alpha1/types_scylladbmanagerclusterregistration.go:
```diff
@@ -0,0 +1,62 @@
+type LocalScyllaDBReference struct {
+	// kind specifies the type of the resource.
+	Kind string `json:"kind"`
+	// name specifies the name of the resource.
```
I thought the `Local` in the struct name is a conventional way of saying this (by similarity to `LocalObjectReference`) - I could add a comment to the entire struct though, if it's not clear enough?

zimnx replied:
struct name isn't visible in `kubectl explain`, while comments are.

rzetelskik replied:
ok, changed comment

rzetelskik commented on pkg/controller/globalscylladbmanager/sync.go:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	smcrMap := map[string]*scyllav1alpha1.ScyllaDBManagerClusterRegistration{}
+	for i := range smcrs {
```
done

rzetelskik commented on pkg/controller/globalscylladbmanager/sync.go:
```diff
@@ -0,0 +1,65 @@
+	}()
+
+	scyllaDBDatacenters, err := gsmc.scyllaDBDatacenterLister.List(globalScyllaDBManagerSelector)
+	if err != nil {
```
done

rzetelskik commented on pkg/controller/globalscylladbmanager/sync.go:
```diff
@@ -0,0 +1,65 @@
+	smcrMap := map[string]*scyllav1alpha1.ScyllaDBManagerClusterRegistration{}
+	for i := range smcrs {
+		smcrMap[smcrs[i].Name] = smcrs[i]
+	}
```
done

rzetelskik commented on pkg/controllertools/observer.go:
```diff
@@ -82,6 +82,10 @@ func (o *Observer) GetGenericHandlers() cache.ResourceEventHandlerFuncs {
 }
 
+func (o *Observer) EventRecorder() record.EventRecorder {
+	return o.eventRecorder
```
squashed

rzetelskik commented on pkg/controller/globalscylladbmanager/sync.go:
```diff
@@ -0,0 +1,65 @@
+func (gsmc *Controller) getScyllaDBManagerClusterRegistrations() (map[string]*scyllav1alpha1.ScyllaDBManagerClusterRegistration, error) {
+	smcrs, err := gsmc.scyllaDBManagerClusterRegistrationLister.List(labels.SelectorFromSet(labels.Set{
+		naming.GlobalScyllaDBManagerLabel: naming.LabelValueTrue,
+	}))
```
done

rzetelskik commented:
> @rzetelskik: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/e2e-gke-parallel-clusterip	[9c7eef1](https://github.com/scylladb/scylla-operator/commit/9c7eef1b8e7ca53ec9ff3ca0056dfa6cff329e45)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel-clusterip/1917218369711378432)	true	`/test e2e-gke-parallel-clusterip`
> ci/prow/e2e-gke-parallel	[9c7eef1](https://github.com/scylladb/scylla-operator/commit/9c7eef1b8e7ca53ec9ff3ca0056dfa6cff329e45)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel/1917218368922849280)	true	`/test e2e-gke-parallel`
> ci/prow/e2e-gke-multi-datacenter-parallel	[9c7eef1](https://github.com/scylladb/scylla-operator/commit/9c7eef1b8e7ca53ec9ff3ca0056dfa6cff329e45)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-multi-datacenter-parallel/1917218371254882304)	true	`/test e2e-gke-multi-datacenter-parallel`
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).

timeouts
/test images
/retest

rzetelskik commented:
> timeouts
> /test images
> /retest

timeouts
/test images
/retest

rzetelskik commented:
> @rzetelskik: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/e2e-gke-parallel-clusterip	[ab9e7eb](https://github.com/scylladb/scylla-operator/commit/ab9e7eba47a55edd06140e1f4340b840bd5e692b)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel-clusterip/1917287479816228864)	true	`/test e2e-gke-parallel-clusterip`
> ci/prow/e2e-gke-parallel	[ab9e7eb](https://github.com/scylladb/scylla-operator/commit/ab9e7eba47a55edd06140e1f4340b840bd5e692b)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel/1917287479551987712)	true	`/test e2e-gke-parallel`
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).

the e2e introduced in this PR failed, I need to look into this

rzetelskik commented:
@zimnx I believe I replied to all your comments, ptal

rzetelskik commented:
> @rzetelskik: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/e2e-gke-parallel-clusterip	[37e3720](https://github.com/scylladb/scylla-operator/commit/37e37207a4e9a0339460814ece40de55175f37d7)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-parallel-clusterip/1917502098874830848)	true	`/test e2e-gke-parallel-clusterip`
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).

timeout
/retest

rzetelskik commented:
Rebased

zimnx commented on test/e2e/set/scylladbdatacenter/scylladbdatacenter_globalmanager.go:
```diff
@@ -0,0 +1,191 @@
+		framework.By("Waiting for ScyllaDBDatacenter to register with global ScyllaDB Manager instance")
+		registrationCtx, registrationCtxCancel := context.WithTimeout(ctx, utils.SyncTimeout)
+		defer registrationCtxCancel()
+		smcr, err := controllerhelpers.WaitForScyllaDBManagerClusterRegistrationState(registrationCtx, nsClient.ScyllaClient().ScyllaV1alpha1().ScyllaDBManagerClusterRegistrations(ns.Name), smcrName, controllerhelpers.WaitForStateOptions{}, utilsv1alpha1.IsScyllaDBManagerClusterRegistrationRolledOut, hasDeletionFinalizer)
```
lets break down this line in a way that conditions we are awaiting for are in separate lines

rzetelskik replied:
done

rzetelskik commented on pkg/controller/globalscylladbmanager/sync_scylladbmanagerclusterregistrations.go:
```diff
@@ -0,0 +1,79 @@
+	} else if globalScyllaDBManagerNamespace.DeletionTimestamp != nil {
+		klog.V(4).InfoS("Global ScyllaDB Manager namespace is being deleted, not creating any ScyllaDBManagerClusterRegistration objects for global ScyllaDB Manager instance.", "Namespace", naming.ScyllaManagerNamespace)
+	} else {
+		for _, sdc := range scyllaDBDatacenters {
+			required, err := makeScyllaDBManagerClusterRegistrationForScyllaDBDatacenter(sdc)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("can't make ScyllaDBManagerClusterRegistration for ScyllaDBDatacenter %q: %w", naming.ObjRef(sdc), err))
+				continue
+			}
+
+			requiredScyllaDBManagerClusterRegistrations[sdc.Namespace] = append(requiredScyllaDBManagerClusterRegistrations[sdc.Namespace], required)
+		}
```
seems purely style-related? It'd just be a wrapper for a for loop?

rzetelskik replied:
done

rzetelskik commented on pkg/controller/globalscylladbmanager/sync_scylladbmanagerclusterregistrations.go:
```diff
@@ -0,0 +1,79 @@
+		return err
+	}
+
+	for _, smcr := range slices.Concat(slices.Collect(maps.Values(requiredScyllaDBManagerClusterRegistrations))...) {
```
it's a 2D slice, concat + unwiding "flattens" it
```
	var a iter.Seq[[]*scyllav1alpha1.ScyllaDBManagerClusterRegistration] = maps.Values(requiredScyllaDBManagerClusterRegistrations)
	var b [][]*scyllav1alpha1.ScyllaDBManagerClusterRegistration = slices.Collect(a)
	var c []*scyllav1alpha1.ScyllaDBManagerClusterRegistration = slices.Concat(b...)
```

rzetelskik commented on pkg/controller/scylladbmanagerclusterregistration/sync.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func (smcrc *Controller) getManagerClient(ctx context.Context, smcr *scyllav1alpha1.ScyllaDBManagerClusterRegistration) (*managerclient.Client, error) {
+	if !isManagedByGlobalScyllaDBManagerInstance(smcr) {
+		return nil, controllertools.NewNonRetriable(fmt.Sprintf("ScyllaDBManagerClusterRegistration %q is not managed by global ScyllaDB Manager instance", naming.ObjRef(smcr)))
+	}
```
we don't currently support any other manager "backend", only the global instance, so this does the only thing possible now for non-supported setups (implicit "managed manager" in that case). Perhaps the error message could be clearer?

mflendrich replied:
I think that @zimnx's point is that this function is in violation of the SRP principle. The "single responsibility" in this case is assembling and returning a client (literally as the function name says).

I second the sentiment of this comment.

Suggestion: make it the caller's responsibility to provide a valid cluster registration (in practice: move the check to the caller level).

rzetelskik replied:
done

rzetelskik commented on pkg/controller/scylladbdatacenter/resource.go:
```diff
@@ -60,17 +60,30 @@ const (
-	maps.Copy(svcLabels, sdc.Labels)
+	maps.Copy(svcLabels, sdcLabels)
 	maps.Copy(svcLabels, naming.ClusterLabels(sdc))
 	svcLabels[naming.ScyllaServiceTypeLabel] = string(naming.ScyllaServiceTypeIdentity)
```
> shouldn't all resources filter those labels/annotations?

I've implemented it so that labels/annotations are filtered out from the propagated labels/annotations - any explicit changes later are not filtered out on purpose (as it's only meant to prevent propagation, not setting the key on an underlying resource altogether).

rzetelskik commented on pkg/controller/scylladbdatacenter/resource.go:
```diff
@@ -2116,3 +2148,19 @@ func MakeUpgradeContextConfigMap(sdc *scyllav1alpha1.ScyllaDBDatacenter, uc *int
+	for _, k := range nonPropagatedLabelKeys {
+		delete(sdcLabels, k)
+	}
+	return sdcLabels
```
I gave it a shot but I don't think it gets any more generic than this. There is `maps.DeleteFunc` which could be used to do
```
maps.DeleteFunc(sdcLabels, func(k, _ string) bool {
  slices.Contains(nonPropagatedLabelKeys, k)
})
```
but that's O(n^2). I think any other approach would be too specific to add it to helpers, so leaving as is.

I extended the `makePropagableAnnotations` func to remove the managed hash annotation though as it was being repeated all over the place.

rzetelskik commented on pkg/controller/scylladbdatacenter/resource_test.go:
```diff
@@ -1688,6 +1688,30 @@ exec scylla-manager-agent \
 			expectedError: nil,
 		},
+		{
+			name: "non-propagated annotations are not propagated",
```
done

rzetelskik commented on pkg/controller/scylladbdatacenter/resource.go:
```diff
@@ -2116,3 +2148,19 @@ func MakeUpgradeContextConfigMap(sdc *scyllav1alpha1.ScyllaDBDatacenter, uc *int
 	}, nil
 }
+
+func getPropagableScyllaDBDatacenterLabels(sdc *scyllav1alpha1.ScyllaDBDatacenter) map[string]string {
```
see https://github.com/scylladb/scylla-operator/pull/2590#discussion_r2091679903

rzetelskik commented on pkg/controller/scylladbmanagerclusterregistration/controller.go:
```diff
@@ -0,0 +1,371 @@
+
+type Controller struct {
+	kubeClient   kubernetes.Interface
+	scyllaClient scyllav1alpha1client.ScyllaV1alpha1Interface
```
done

rzetelskik commented on pkg/controller/globalscylladbmanager/controller.go:
```diff
@@ -0,0 +1,316 @@
+	smcr := obj.(*scyllav1alpha1.ScyllaDBManagerClusterRegistration)
+
+	if !isManagedByGlobalScyllaDBManagerInstance(smcr) {
+		klog.V(5).InfoS("Not enqueueing ScyllaDBManagerClusterRegistration not owned by global ScyllaDB Manager", "ScyllaDBManagerClusterRegistration", klog.KObj(smcr), "RV", smcr.ResourceVersion)
```
done

rzetelskik commented on test/e2e/utils/verification/scyllacluster/cql.go:
```diff
@@ -27,22 +28,11 @@ func WaitForFullMultiDCQuorum(ctx context.Context, dcClientMap map[string]corev1
 }
 
 func VerifyCQLData(ctx context.Context, di *utils.DataInserter) {
-	err := di.AwaitSchemaAgreement(ctx)
```
moved to test/e2e/utils/verification/cql.go

rzetelskik commented on pkg/controller/scylladbmanagerclusterregistration/sync_finalizer.go:
```diff
@@ -0,0 +1,101 @@
+	err = smcrc.deleteScyllaDBManagerCluster(ctx, smcr)
+	if err != nil {
+		return progressingConditions, fmt.Errorf("can't delete ScyllaDB Manager cluster: %w", err)
+	}
```
done

rzetelskik commented:
@zimnx I addressed all of your comments, ptal https://github.com/scylladb/scylla-operator/compare/4fab497c5cfe62dbd5d62ba8037f81eb76f19251..b88c3c9e0eacd54cdc3d1d9ed10071ebb45f5a9f

zimnx commented on pkg/controller/scylladbdatacenter/resource.go:
```diff
@@ -316,22 +317,21 @@ func StatefulSetForRack(rack scyllav1alpha1.RackSpec, sdc *scyllav1alpha1.Scylla
 	rackLabels := map[string]string{}
-	maps.Copy(rackLabels, sdc.Labels)
 	maps.Copy(rackLabels, requiredLabels)
+	maps.Copy(rackLabels, sdcLabels)
```
sdc.Labels and requiredLabels were swapped, which may be incorrect, as required ones should be added at the end to win potential collisions. Any reason why this is changed?

zimnx reviewed (commented):
Please reduce changes in resource_test in scylladbdatacenter controller to minimum, with current diff it's impossible to tell what was changed.

rzetelskik commented:
> Please reduce changes in resource_test in scylladbdatacenter controller to minimum, with current diff it's impossible to tell what was changed.

Sent https://github.com/scylladb/scylla-operator/pull/2645, I'll rebase when it's merged. Only the new test cases should be included in this PR then.
/hold

rzetelskik commented on pkg/controller/scylladbdatacenter/resource.go:
```diff
@@ -2093,26 +2059,44 @@ func MakeUpgradeContextConfigMap(sdc *scyllav1alpha1.ScyllaDBDatacenter, uc *int
+		delete(annotations, k)
+	}
+
+	// As ScyllaDBDatacenter may be a managed object (when user is using scyllav1.ScyllaCluster API), its managed
+	// hash shouldn't be propagated into dependency objects to avoid triggering unnecessary double rollouts.
```
good idea, fixed

rzetelskik commented on pkg/controller/scylladbdatacenter/resource.go:
```diff
@@ -60,20 +60,24 @@ const (
-	svcAnnotations := map[string]string{}
+	// Label keys excluded from propagation to underlying resources.
+	nonPropagatedLabelKeys = []string{
+		naming.GlobalScyllaDBManagerRegistrationLabel,
```
done

rzetelskik commented:
rebased
/hold cancel

rzetelskik commented:
I believe all remaining comments were addressed. On top of that, I:
- Fixed a bug where using `maps.Clone` in creation of propagated labels could cause a nil map access.
- Disabled creation of the default repair task on cluster creation. This is to prevent name collisions with ScyllaDBManagerTasks (to be introduced in subsequent PRs). It's also aligned with existing behaviour for ScyllaClusters.

/cc zimnx

rzetelskik commented:
/retest

rzetelskik commented:
- Removed "global" timeout from new E2E specs

rzetelskik commented:
> @rzetelskik: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name | Commit | Details | Required | Rerun command
> --- | --- | --- | --- | ---
> ci/prow/e2e-gke-multi-datacenter-parallel | 236097fa7455864e3dead8327453ed68212d0fa3 | [link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-multi-datacenter-parallel/1927450758425350144) | true | `/test e2e-gke-multi-datacenter-parallel`
> 
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).
> 
> <details>
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
> </details>
> <!-- test report -->

timeout
/retest

mflendrich commented on pkg/controller/scylladbdatacenter/resource.go:
```diff
@@ -2116,3 +2087,21 @@ func MakeUpgradeContextConfigMap(sdc *scyllav1alpha1.ScyllaDBDatacenter, uc *int
 	}, nil
 }
+
+func makePropagableLabels(sdc *scyllav1alpha1.ScyllaDBDatacenter) map[string]string {
+	labels := map[string]string{}
+	maps.Copy(labels, sdc.Labels)
+	for _, k := range nonPropagatedLabelKeys {
+		delete(labels, k)
+	}
+	return labels
+}
+
+func makePropagableAnnotations(sdc *scyllav1alpha1.ScyllaDBDatacenter) map[string]string {
+	annotations := map[string]string{}
+	maps.Copy(annotations, sdc.Annotations)
+	for _, k := range nonPropagatedAnnotationKeys {
+		delete(annotations, k)
+	}
+	return annotations
+}
```
Duplicate logic (DRY) - suggestion: refactor into a generic function and inject the values as appropriate:

```go
mapCopyExcludingKeys(sdc.Annotations, nonPropagatedAnnotationKeys)
mapCopyExcludingKeys(sdc.Labels, nonPropagatedLabelKeys)
```

rzetelskik replied:
done

mflendrich replied:
> I'll amend the code to just inline the stdlib functions then.

This misses the point of my suggestion, which is:

- maintaining the fact that "creating a copy with elements filtered out" is a single function call
- not duplicating logic if not necessary.

I would judge that keeping a function as generic as necessary for the reasonable set of call sites is enough - unless we're fixing this across a wide range of applications/places, I think that operating on `map[string]string` and `[]string` should be just fine.

Therefore my suggestion is to either:
- apply the original suggestion
- keep the original as it was

I am under impression that inlining the stdlib functions satisfies no one (you included).

mflendrich commented on test/e2e/utils/helpers.go:
```diff
@@ -285,24 +286,7 @@ func GetStatefulSetsForScyllaCluster(ctx context.Context, client appv1client.App
 }
 
 func GetPodsForStatefulSet(ctx context.Context, client corev1client.CoreV1Interface, sts *appsv1.StatefulSet) (map[string]*corev1.Pod, error) {
```
this function is now a trivial wrapper - why not delete it altogether?

rzetelskik replied:
The intention was:
- not duplicating the code between v1 and v1alpha1 utils
- not having to modify all existing tests for v1
- tests being able to use a single import (v1 utils xor v1alpha1 utils)

mflendrich reviewed (commented):
https://github.com/scylladb/scylla-operator/pull/2590#discussion_r2111909290

mflendrich reviewed (approved):
Thanks!

/lgtm

rzetelskik commented:
> @rzetelskik: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name | Commit | Details | Required | Rerun command
> --- | --- | --- | --- | ---
> ci/prow/e2e-gke-serial | a83fcfd024cb22978885262e830969d4ae794880 | [link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2590/pull-scylla-operator-master-e2e-gke-serial/1927729873019736064) | true | `/test e2e-gke-serial`
> 
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=2590). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).
> 
> <details>
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
> </details>
> <!-- test report -->

apiserver
/retest
